### ISSUE TITLE
Multithreaded backpropagation with custom autograd.Functions

### ISSUE THREAD

**Issue by cannon** [pytorch/pytorch#25514]
## 🚀 Feature
By using DataParallel, the engine creates a thread for each GPU. It should do this for backpropagation too, including within python-implemented autograd.Functions.

## Motivation
We're trying to implement synchronized batchnorm for one machine/one "main" process using an interface like the traditional DataParallel. To synchronize the forward pass (computation of the mean+variance) we use a threading.Barrier to sync the threads for each GPU.

As it turns out, torch's batchnorm uses a custom backwards function which gives a different gradient than you'd get just from doing a linear transform with the mean and variance: https://github.com/pytorch/pytorch/blob/v1.1.0/aten/src/ATen/native/Normalization.cpp

To implement a SyncBatchNorm that behaves identically on multiple GPUs as torch's batchnorm does on one GPU it looks like we'd need to synchronize things like the mean of grad_output across all GPUs. But we can't get our .backward to be entered by multiple threads simultaneously to pull this off.

## Additional context
At first we were concerned that torch might not multithread backprop at all, which would be a huge slowdown for multi-GPU training, but it looks like it does for C implemented functions only? I can't find any clear documentation.

**Comment by soumith** [comment]
@cannon have you considered using `DistributedDataParallel` even for single machine multi-gpu case? In that case, you just need to all reduce once, instead of the `DataParallel` model which does get more complicated. We even have SyncBatchNorm implemented for the DistributedDataParallel case: https://pytorch.org/docs/stable/nn.html?highlight=syncbatchnorm#torch.nn.SyncBatchNorm
